tasmota/berry/lorawan/decoders/LwDecode.be

```diff
@@ -138,6 +138,116 @@ end
+  end
+
+  #- this method displays the web page -#
+  def pageLoRaWAN()
+    if !webserver.check_privileged_access() return nil end
+
+    var inode=1
+    var cmdArg  
+    if webserver.has_arg('save')
+     inode = webserver.arg('node')
+     tasmota.cmd('LoRaWanAppKey'+inode+' '+ webserver.arg('ak'),true)
+     cmdArg = webserver.arg('dc')
+     if !cmdArg cmdArg='"' end
+     tasmota.cmd('LoRaWanDecoder'+inode+' '+cmdArg,true)
+     cmdArg = webserver.arg('an')
+     if !cmdArg cmdArg='"' end
+     tasmota.cmd('LoRaWanName'+inode+' '+cmdArg,true)
+    end
+
+    webserver.content_start("LoRaWAN")           #- title of the web page -#
+    webserver.content_send_style()               #- send standard Tasmota styles -#
+    webserver.content_send(
+     "<style>.inactive{background:#10537c;}.active{background:#1fa3ec;}</style>"
+     "<h2>LoRaWAN End Devices</h2>"
+     "<script>"
+     "function selNode(n){"
+     " var i;var d=4;"
+     " var e=document.getElementById('n'+n);"
+     " var o=document.getElementsByClassName('button active');"
+     " if(o.length){"
+     "  for(i=0;i<o.length;i++){"
+     "  o[i].classList.add('inactive');"
+     "  o[i].classList.remove('active');"
+     "  }"
+     " }"
+     " e.classList.add('active');"
+     " for(i=1;i<=16;i++){"
+     "  document.getElementById('nd'+i).style.display=(i==n)?'block':'none';"
+     " }"
+     "}"
+     "window.onload = function(){selNode("+str(inode)+");};"
+     "</script>")
+
+    var arg, appKey, decoder, name
+    var hintAK='32 character Application Key'
+    var hintDecoder='Decoder file, ending in .be'
+    var hintAN='Device name for MQTT messages'
+    webserver.content_send("<table><tr>")
+    for node:1..8
+        self.sendNodeButton(node)
+    end
+    webserver.content_send("<tr>")
+    for node:9..16
+        self.sendNodeButton(node)
+    end
+    webserver.content_send("</table>")
+    for node:1..16
+     arg='LoRaWanAppKey' + str(node)
+     appKey=tasmota.cmd(arg,true).find(arg)
+     arg='LoRaWanName' + str(node)
+     name=tasmota.cmd(arg,true).find(arg)
+     arg='LoRaWanDecoder' + str(node)
+     decoder=tasmota.cmd(arg,true).find(arg)
+     webserver.content_send(
+      "<div id='nd"+str(node)+"' style='display:none'>"
+      "<fieldset>"
+      "<legend><b>&nbsp;End Device "+str(node)+"&nbsp;</b></legend>"
+      "<form action='' method='post'>"
+      "<p>Application Key"
+      "<input title='"+hintAK+"' pattern='[A-Fa-f0-9]{32}' id='ak' minlength='32' maxlength='32' required='' placeholder='"+hintAK+"' value='"+appKey+ "' name='ak' style='font-size:smaller'>"
+      "</p>"
+      "<p></p>"
+      "<p>Device Name"
+      "<input id='an' placeholder='"+hintAN+"' value='"+name+ "' name='an'>"
+      "</p>"
+      "<p></p>"
+      "<p>Decoder File"
+      "<input title='"+hintDecoder+"' id='dc'  placeholder='"+hintDecoder+"' value='"+decoder+"' name='dc'>"
```

Comment:
Just a quick comment, it is very costly to concatenate many strings.

Prefer an f-string construct that allocates only one buffer for the result:
```
f"content"
 "content following on next line"
 "with inserts: {hintAN}"
```